### PR TITLE
GIT-6: Clone non-public repositories

### PR DESCRIPTION
**Pull request by Haxsen**
[Issue link](https://jira.xwiki.org/browse/GIT-6)

This PR is mainly for correction and learning purposes.

Main addition:
`getRepository` method (overload) which takes extra params: username and token. Using the credentials, it clones the private repository.
Note: `GitManagerNew` is probably a bad name or implementation, so I would like your help.

Summary:
Due to revapi checks failing, only way I found was to:
1. Create a new interface `GitManagerNew` which extends original `GitManager`.
2. Use the new `GitManagerNew` in `DefaultGitManager` and `GitScriptService` classes.
3. Create the `getRepository` method (overloaded), alternative to the original, which takes username+token as the credentials and use them to clone a private repository (if that user has access to it).

Before the commit "Revert commit", those commits are the other way which work as well with `revapi-check`. But worse than the one I used now (probably).

Your review is highly appreciated, big thanks for spending your valuable time here 😄 

* Before:
PrivateGitBefore

* After:
PrivateGitAfter

### CODE REVIEW

**Comment by Haxsen** [comment]
Forgot to add screenshots, doing now.

File: src/main/java/org/xwiki/git/GitManagerNew.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * Provides services to access a Private Git repository.
+ *
+ * @version $Id$
+ * @since 12.3RC1
+ */
+@Unstable
+@Role
+public interface GitManagerNew extends GitManager
```
Comment (vmassol) [inline src/main/java/org/xwiki/git/GitManagerNew.java:34]: bad name....

Comment (vmassol) [reply src/main/java/org/xwiki/git/GitManagerNew.java:34]: Should not be a new interface

File: src/main/java/org/xwiki/git/internal/DefaultGitManager.java [hunk 2]
```diff
@@ -101,6 +101,41 @@ public Repository getRepository(String repositoryURI, String localDirectoryName)
...
+                        .setCredentialsProvider(new UsernamePasswordCredentialsProvider(username, token))
+                        .setDirectory(localDirectory)
+                        .setURI(repositoryURI)
+                        .call();
+            }
+        } catch (Exception e) {
+            throw new RuntimeException(String.format("Failed to execute Git command in [%s]", gitDirectory), e);
+        }
+
+        return repository;
```
Comment (vmassol) [inline src/main/java/org/xwiki/git/internal/DefaultGitManager.java:136]: Doesn't look good. Seems you copy pasted existing code, increasing the technical debt.

**Comment by vmassol** [comment]
You need to provide a real PR, because we cannot apply this. Thanks.

**Comment by Haxsen** [comment]
Ok @vmassol working on the changes requested by your review.
Thanks.

File: src/main/java/org/xwiki/git/GitManagerNew.java (hunk 1 above)
Comment (Haxsen) [reply src/main/java/org/xwiki/git/GitManagerNew.java:34]: If I create the method inside GitManager, I get error on revapi-check. Finding another way.

File: src/main/java/org/xwiki/git/internal/DefaultGitManager.java (hunk 2 above)
Comment (Haxsen) [reply src/main/java/org/xwiki/git/internal/DefaultGitManager.java:136]: Yep, I am combining the common lines now.

File: src/main/java/org/xwiki/git/GitManagerNew.java (hunk 1 above)
Comment (vmassol) [reply src/main/java/org/xwiki/git/GitManagerNew.java:34]: Revapi is correct. You must not break backward compatibility! :) Please research it a bit, it should be easy to find the answer (if I tell you then I'm doing all the job and it won't prove your ability to research a topic and think about finding a solution). Thanks!

Comment (Haxsen) [reply src/main/java/org/xwiki/git/GitManagerNew.java:34]: Got it, thanks. I will keep trying until you're satisfied :)

**Pull request closed without merging**